Fixed #30581 -- Added support for Meta.constraints validation.

Gagaro opened the pull request:
https://code.djangoproject.com/ticket/30581

Still needed:

- [x] Update `Model.full_clean` to take advantage of this new `Constraint.validate`
- [x] Documentation
- [x] More tests (`Additional tests around exclude usage would also be appreciated.`)

About `Model.full_clean`, as it have a parameter `validate_unique`, should we still have a special case for `UniqueConstraint`, excluding them from `Model.full_clean` to only have them checked in `Model.validate_unique`?

felixxm commented on django/db/backends/base/operations.py:
```diff
@@ -53,6 +53,9 @@ class BaseDatabaseOperations:
     explain_prefix = None
 
+    # FROM statement for SELECT queries against no tables.
+    empty_from_sql = None
```
You can reuse `bare_select_suffix`.

Gagaro commented:
Should the Q.check method be part of the public API (and thus documented)?

charettes commented:
@Gagaro sorry for the late reply. I think we should focus on the constraint validation work here and defer the decision of making this method public and documented.

charettes commented on django/db/models/constraints.py:
```diff
@@ -64,6 +78,19 @@ def create_sql(self, model, schema_editor):
+            against[field.name] = Value(value, field)
+        try:
+            if not self.check.check(against):
+                raise ValidationError(f'Constraint {self.name} is violated')
```
We should make use of `self.message`.

charettes commented on django/db/models/constraints.py:
```diff
@@ -1,17 +1,28 @@
-    def __init__(self, name):
+    message = _('A constraint failed.')
+
+    def __init__(self, name, message=None):
```
Wonder if we should have a more explicit name here, `message` makes sense in the case of `Validator` but it seems a bit ambiguous in the case of constraint. Thoughts @django/triage-review?

pope1ni replied:
As far as I can see `message` is always `None`. It doesn't seem to be used. Can it just be removed?

charettes replied:
The point of it is for the `ValidationErrror` raised to use this message just like `Validator(message)` is used

https://github.com/django/django/blob/910ecd1b8df7678f45c3d507dde6bcb1faafa243/django/core/validators.py#L27-L51

felixxm replied:
We should add a test for passing a `message` to the constraint.

Gagaro replied:
I added tests on `BaseConstraint.get_violation_error_message`. Should we also test passing custom messages for the implemented constraints (`CheckConstraint`, ...) as they only pass it through `super`?

charettes commented on django/db/models/query_utils.py:
```diff
@@ -78,6 +79,26 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
         query.promote_joins(joins)
         return clause
 
+    def check(self, against, using=DEFAULT_DB_ALIAS):
+        """
+        Do a database query to check if the expressions of the Q instance
+        matches against the `against` expressions.
+        """
+        # Avoid circular imports
+        from django.db.models.sql import Query
+        from django.db.models.sql.constants import SINGLE
+        from django.db.models import Value, ExpressionWrapper, BooleanField
+
+        query = Query(None)
+        for name, value in against.items():
+            if not hasattr(value, 'resolve_expression'):
+                value = Value(value)
+            query.add_annotation(value, name, select=False)
+        # This will raise a FieldError if a needed field is missing from against
+        query.add_annotation(ExpressionWrapper(self, BooleanField()), '_check')
+        compiler = query.get_compiler(using=using)
+        return bool(compiler.execute_sql(SINGLE)[0])
+
```
The discussion about making this method public makes me wonder if we should not make it a private method of `BaseConstraint` for the time being.

carltongibson replied:
@charettes I'd imagined this being user-accessible (public). Originally, I'd imagined folks implementing this in Python (`return x > 18` sort of thing). Your suggestion to go via the DB is significantly better (more robust) but I'd still imagine folks wanting to implement constraints for which a simple Python check would be enough. Documenting the hook would allow that... 🤔 (Discuss :)

charettes replied:
> I'd still imagine folks wanting to implement constraints for which a simple Python check would be enough. 

@carltongibson in this case we need to document `Constraint.validate` and not necessarily `Q.check` don't we?

carltongibson replied:
Yes, that covers my use-case. (Sorry)

charettes replied:
No worries, just wanted to make sure we were on the same line.

charettes reviewed (commented):
Thanks for your work so far!

I think we should also add a commit that makes `Model.full_clean` accept a `validate_constraint=True` kwarg and make sure constraint validation is performed when provided otherwise ticket-30581 won't actually be solved by this PR.

Additional tests around `exclude` usage would also be appreciated.

Gagaro commented:
Thanks for the review.

About `Model.full_clean`, where should the `UniqueConstraints` be validated?

Right now they are in `validate_unique`. Should we let them here (and use the new `BaseConstraint.validate` instead of what is currently done, including the `UniqueConstraints` with conditions) and remove them from the new `validate_constraints` as a special case, or remove them from `validate_unique` and check them with the others constraints?

charettes commented:
I think that moving all constraint validation to a dedicated `validate_constraint` method, meaning removing constraint handling logic from `Model._get_unique_checks`, would make the most sense here.

Ideally `UniqueConstraint` validation error message would still make use of `Model.unique_error_message` if it's not conditional. I think that'll be necessary in order to have existing regression tests passing.

Gagaro commented:
I added the validate_constraints to `Model.full_clean` and added more tests for `exclude` (which allowed me to find some bugs I fixed).

There is something that is not exactly done the same way with `UniqueConstraint`, I don't have the actual model class to pass to `unique_error_message` in [validate](https://github.com/Gagaro/django/blob/ticket-30581/django/db/models/constraints.py#L326).

Should we add a special case for `UniqueConstraint` and pass the model class to `UniqueConstraints.validate` [here](https://github.com/Gagaro/django/blob/ticket-30581/django/db/models/base.py#L1209-L1212)?

charettes commented:
Can't the model class be inferred from `instance`?

Gagaro commented:
Yes but it duplicates the code to do it. I added `Model.get_constraints` to avoid that but it is still called at least twice.

Maybe it should be a `cached_property`?

Gagaro commented:
Hi, I tried to add documentation but I wasn't sure how to format or add the new information. It's a first step for it at least :smiley: .

Gagaro commented:
I don't know why but a table is missing in MySQL?

`MySQLdb._exceptions.ProgrammingError: (1146, "Table 'test_django_39_pull-requests-bionic.validation_uniqueconstraintconditionproduct' doesn't exist")`

Gagaro commented on django/db/models/constraints.py:
```diff
@@ -253,3 +282,57 @@ def deconstruct(self):
+                if not (~self.condition | ~Exists(queryset)).check(against):
+                    raise ValidationError(self.get_message())
+            except FieldError:
+                return None
```
I agree, I removed the `None`.

Gagaro commented:
Finally had the time to look at the MySQL issue. It was because a model was excluded (`required_db_features = {'supports_partial_indexes'}`) but the tests were not skipped.

I also squashed all the commits to ease the rebase.

Gagaro commented:
Hi @charettes , what are the next steps for this PR?

Are there enough tests / documentation?

charettes reviewed (commented):
Things are looking great, thanks for your continued efforts here.

Apart from a more proper commit breakdown I think the patch in its current state makes sense and addresses ticket-30581. It feels like it would be a nice feature for Django 4.1 which I've seen you've targeted here.

@django/triage-review a second ping from me tonight but I'd appreciate your input on this one as it was heavily driven by @Gagaro and myself and having a few more eyes look at the suggested implementation would certainly be beneficial.

kezabelle commented:
So, my familiarity with this bit of the ORM is ... less than stellar; as such take my comments with a grain of salt.

The only thing that I am really scratching my head (so far) about when _scanning_ the diff, is the following conditions split across a  few places:
```
if exclude and field.name in exclude:
    continue
[...]
if isinstance(expr, F) and expr.name in exclude:
    return
[...]
if exclude and expression.name in exclude:
    return
```

Evidence would suggest from the various tests passing that these are each individually and collectively correct, but I cannot tell _why_ it sometimes returns early and sometimes continues iterating for what look to be basically 'the same' kind of test (If I'm reading it correctly!). 
Personally I'd be inclined to have a comment on the ones which return early explaining why it's fine to stop exhaustively checking entirely at that point; though if it's clear enough to everyone else and I'm just not grokking the whole thing, that's also a fine outcome :)

kezabelle commented on django/db/models/constraints.py:
```diff
@@ -253,3 +282,57 @@ def deconstruct(self):
+                        if isinstance(expr, F) and expr.name in exclude:
+                            return
+            queryset = queryset.filter(*expressions)
+        if not instance._state.adding and instance.pk is not None:
+            queryset = queryset.exclude(pk=instance.pk)
```
Same comment as prior, about using `_get_pk_val` instead. Disregard this if that comment was irrelevant.

charettes replied:
It was definitely relevant, let's continue the discussion over there.

kezabelle commented on docs/releases/4.1.txt:
```diff
@@ -219,6 +219,9 @@ Models
   in order to reduce the number of failed requests, e.g. after database server
   restart.
+* New :meth:`BaseConstraint.validate(instance, exclude=None, using=DEFAULT_DB_ALIAS) <.BaseConstraint.validate>`
+  method returns whether the instance validate the constraint or not. This needs
+  to be implemented for each subclasses.
```
I _think_ `validate` here ought to be `validates` or `validated`?
`each subclasses` should be `each subclass` or `all subclasses` I believe.

charettes replied:
Agreed, we also still need to determine if we actually want to force all subclasses to implement this form of validation.

charettes commented on django/contrib/postgres/constraints.py:
```diff
@@ -178,3 +180,40 @@ def __repr__(self):
+            lookup.postgres_operator = operator
+            lookups.append(lookup)
+        queryset = queryset.filter(*lookups)
+        if not instance._state.adding and instance.pk is not None:
+            queryset = queryset.exclude(pk=instance.pk)
```
That's a nice catch @kezabelle definitely something the logic should take into account that we should add tests for like 53fb45c6d8 did for ticket-17615.

The `validate` logic will need to know which model/table the constraint is _bound_ to in order to properly support mutli-table inheritance which it doesn't do at all right now.

For example, even putting the ticket-17615 issue aside, in the case of

```python
constraint = UniqueConstraint('x', 'y', name='unique_position')

class Position(models.Model):
    x = models.IntegerField()
    y = models.IntegerField()

    class Meta:
        contraints = [constraint]


class ImportantPosition(Positition):
    pass

Position.objects.create(x=1, y=2)
constraint.validate(ImportantPosition(x=1, y=2))
```

No validation error would be raised as we'd attempt

```sql
SELECT 1
FROM important_position
JOIN position ON (important_position.position_ptr_id = position_id)
WHERE x = 1 AND y 2
```

And since there's no existing `important_position` row the violation wouldn't be caught. What we need to perform here is

```sql
SELECT 1
FROM position
WHERE x = 1 AND y 2
``` 

Even when a child class instance is provided.

Ideally we'd bind `Constraint` classes to the model they are associated to at definition time (e.g. by setting `.model` on them like we do `Field` instances at `Option.contribute_to_class` time) so `.validate` can do `queryset = self.model._default_manager.using(using)` and `instance_pk = self._get_pk_val(self.model._meta)` but that will make testing of `Constraint` in isolation awkward (think of all the `Field.contribute_to_class` class in tests).

An alternative might be to change the `validate` signature to `(model: Type[Model], instance: Model, ...)` which is in line with `create_sql` and friends signature.

In all cases we need:
1. Addition regression test for perfoming validation on an instance of a model inheriting a constraint from a MTI parent
2. Adjustments of the validation logic to perform queries scoped to the model where the constraint was defined and not against the model of the provided instance.

charettes reviewed (commented):
@Gagaro, @kezabelle raised an important problem regarding inheritance and I tried to provide leads as to how we could address it. Let me know if that makes sense to you!

Gagaro commented on django/db/models/constraints.py:
```diff
@@ -25,6 +34,12 @@ def create_sql(self, model, schema_editor):
         raise NotImplementedError('This method must be implemented by a subclass.')
 
+    def validate(self, instance, exclude=None, using=DEFAULT_DB_ALIAS):
+        raise NotImplementedError('This method must be implemented by a subclass.')
```
Should we show a warning then, or do we plan on letting validate optional?

Gagaro commented:
Yes I see the issue. I'll take a look at it.

Gagaro commented on django/db/models/base.py:
```diff
@@ -1272,6 +1286,16 @@ def full_clean(self, exclude=None, validate_unique=True):
+        # Run constraints checks, but only for fields that passed validation.
+        if validate_constraints:
+            for name in errors:
+                if name != NON_FIELD_ERRORS and name not in exclude:
+                    exclude.append(name)
+            try:
+                self.validate_constraints(exclude=exclude)
```
Did you mean `Model.clean_fields(exclude)`? It could also be a set there.

Gagaro commented:
> The only thing that I am really scratching my head (so far) about when _scanning_ the diff, is the following conditions split across a few places:
> 
> ```
> if exclude and field.name in exclude:
>     continue
> [...]
> if isinstance(expr, F) and expr.name in exclude:
>     return
> [...]
> if exclude and expression.name in exclude:
>     return
> ```

They are in different `validate()` implementations (`CheckConstraint.validate()`, `UniqueConstraint.validate()`, ...) and in different branches of the same condition. Exclusion is only checked once per `validate()` call if I'm not mistaken:

```python
 if isinstance(expression, F):
    if exclude and expression.name in exclude:
        return
    [...]
else:
    [...]
    if isinstance(expr, F) and expr.name in exclude:
        return
```

Gagaro commented:
I changed the `validate()` signature to have an additional `model` parameter.

Having the model attached to the constraint would be more convenient IMO but a refactoring of all the other methods (`create_sql` and so on) should be done at the same time to profit from it. It seems out of the scope of this PR so I preferred the other method.

I still have the commits split to do.

Gagaro commented:
I have an issue with `OpClass`. `opclasses` in `ExclusionConstraint` was deperecated in favor of the new `OpClass`. However, `OpClass` should only be used during the index creation. But as it is implemented, it is also rendered in the query in `validate`:

```python
class OpClass(Func):
    template = '%(expressions)s %(name)s'

    def __init__(self, expression, name):
        super().__init__(expression, name=name)
```

So the following constraints:

```python
ExclusionConstraint(
    name='exclude_overlapping_reservations_custom',
    expressions=[
        (
            OpClass(TsTzRange('start', 'end', RangeBoundary()), 'range_ops'),
            RangeOperators.OVERLAPS
        ),
        (OpClass('room', 'gist_int4_ops'), RangeOperators.EQUAL)
    ]
)
```

Create a query like:

```sql
WHERE (TSTZRANGE(U0."start", U0."end", '[)') range_ops && (TSTZRANGE(2018-06-24 00:00:00, 2018-06-28 00:00:00, '[)') range_ops) AND U0."room_id" gist_int4_ops = (1 gist_int4_ops))
```

Notice the `range_ops` and `gist_int4_ops` inserted into the query.

I'm not sure where to fix that:

* In the validate , by removing the `OpClass` from the expressions?
* In OpClass directly? But how?
* Also, can we have the issue with `OrderBy` and `Collate` (which are the others two index wrappers)? `Collate` might be important  and I'm not sure it would cause an issue. Is it possible to find an `OrderBy` in constraints?

Thanks for the help.

charettes commented:
I think that it should be _ok_ to have `ExclusionConstraint` special case `OpClass` expressions somehow for now? What about making `PostgresOperatorLookup.__init__` strip classes from its provided lhs and rhs? It seems like it would nicely contain this _operator_ logic? Maybe a dedicated `PostgresOperatorLookup.from_constraint(lhs, rhs, operator)` would make sense to avoid the manual `.postgres_operator` assignment?

`Collate` is necessary and must not be replaced. e.g. imagine creating a unique constraint for usernames with a case-insensitive collation, you'd want the user input to be collated using the same rules.

As for `OrderBy` I don't think you can create a constraint out of them so I wouldn't worry. Sure you can create an ordered index but it doesn't make sense in the context of a constraint.

charettes commented on django/db/models/lookups.py:
```diff
@@ -297,6 +297,19 @@ def as_postgresql(self, compiler, connection):
         params = tuple(lhs_params) + tuple(rhs_params)
         return '%s %s %s' % (lhs, self.postgres_operator, rhs), params
 
+    @classmethod
+    def from_constraint(cls, lhs, rhs, operator):
+        from django.contrib.postgres.indexes import OpClass
+
+        # We need to get rid of OpClass because it only has sense during the constraint creation
+        args = [
+            expr.get_source_expressions()[0] if isinstance(expr, OpClass) else expr
+            for expr in (lhs, rhs)
+        ]
+        instance = cls(*args)
+        instance.postgres_operator = operator
+        return instance
+
```
This should live on `PostgresOperatorLookup` and not `Lookup` itself; both `OpClass` and `ExclusionConstraint` are `django.contrib.postgres` construct.

charettes replied:
NVM, just realized this is actually `PostgresOperatorLookup` 🤦 I was under the impression that it lived in `django.contrib.postgres.lookups` but it seems to be relied on by `JSONField`.

Maybe that it would be better to confine the logic in `ExclusionConstraint.validate` then 🤷

Gagaro replied:
Shouldn't `PostgresOperatorLookup` also be in `django.contrib.postgres` then?

charettes replied:
Possibly but this patch is already touching a lot of areas so I'd suggest moving this whole `from_constraint` logic back to `ExclusionConstraint.validate` instead.

Sorry for the back and forth here.

ngnpope replied:
> Shouldn't `PostgresOperatorLookup` also be in `django.contrib.postgres` then?

No, because then `django.db` will depend on `django.contrib`.

Also, it looks like JSON support is being baked into the next version of SQLite, and it'll support some PostgreSQL-style operators too. So `PostgresOperatorLookup` could become more generic.

charettes commented on tests/validation/models.py:
```diff
@@ -136,3 +136,52 @@ class Meta:
         ]
+
+
+class Product(models.Model):
+    price = models.IntegerField(null=True)
+    discounted_price = models.IntegerField(null=True)
+
+    class Meta:
+        required_db_features = {
+            'supports_table_check_constraints',
+        }
+        constraints = [
+            models.CheckConstraint(
+                check=models.Q(price__gt=models.F('discounted_price')),
+                name='price_gt_discounted_price_validation',
+            ),
+        ]
+
+
+class ChildProduct(Product):
+    pass
```
I think this model is causing the MySQL failures.

Since `Product` cannot be created on MySQL due to not having the `supports_table_check_constraints` feature then `ChildProduct` cannot be created either since MTI requires a foreign key (hence the `Cannot add foreign key constraint` error).

```suggestion
class Product(models.Model):
    price = models.IntegerField(null=True)
    discounted_price = models.IntegerField(null=True)

    class Meta:
        required_db_features = {
            'supports_table_check_constraints',
        }
        constraints = [
            models.CheckConstraint(
                check=models.Q(price__gt=models.F('discounted_price')),
                name='price_gt_discounted_price_validation',
            ),
        ]


class ChildProduct(Product):
     class Meta:
        required_db_features = {
            'supports_table_check_constraints',
        }
```

charettes replied:
I think this model is causing the MySQL failures.

Since `Product` cannot be created on MySQL due to not having the `supports_table_check_constraints` feature then `ChildProduct` cannot be created either since MTI requires a foreign key (hence the `Cannot add foreign key constraint` error).

```suggestion
class ChildProduct(Product):
     class Meta:
        required_db_features = {
            'supports_table_check_constraints',
        }
```

Gagaro commented:
Also I realized that `OpClass` is not in the `__all__` list (https://github.com/django/django/blob/main/django/contrib/postgres/indexes.py#L5-L8), this should probably be addressed in a new ticket?

Gagaro commented:
I moved the logic from `PostgresOperatorLookup` to `ExclusionConstraint` and split the commit.

That made me realize we actually don't have any test for `Q.check`. I'll add some later today or this week.

Gagaro commented:
Rebase after the black PR done :ok_hand: .

Gagaro commented:
Hi @charettes , what's next on that PR? Thanks.

felixxm commented on django/db/models/sql/query.py:
```diff
@@ -46,6 +46,8 @@
 
 
 def get_field_names_from_opts(opts):
+    if opts is None:
+        return set()
```
This is only tested in `constraints.tests.CheckConstraintTests.test_validate` and `constraints.tests.UniqueConstraintTests.test_validate_expression_condition`. Can we add a regression test in the `FromEmptyQuery`?

felixxm commented on django/db/models/sql/compiler.py:
```diff
@@ -797,6 +801,8 @@ def get_default_columns(self, start_alias=None, opts=None, from_parent=None):
         if opts is None:
             opts = self.query.get_meta()
+            if opts is None:
+                return result
```
We can use the walrus operator here:
```python
if (opts := self.query.get_meta()) is None:
    return result
```

felixxm commented on django/db/models/sql/query.py:
```diff
@@ -1083,8 +1088,9 @@ def join_parent_model(self, opts, model, alias, seen):
+        allow_join = self.model is not None
         annotation = annotation.resolve_expression(
-            self, allow_joins=True, reuse=None, summarize=is_summary
+            self, allow_joins=allow_join, reuse=None, summarize=is_summary
```
This is untested.

Gagaro replied:
I am not sure how to test that one, as the only place I could find the allow_joins used required having a model to go there.

I also don't remember why it was modified. Should I just mock the resolve_expression and check `allow_joins` value?

Gagaro replied:
I tried setting it to True as it was before and no test failed. Maybe we should juste revert this part. @charettes , do you remember why it was modified?

felixxm replied:
In all cases `resolve_expression()` will raise `FieldError` when no model is specified, before trying to join anything. IMO we can revert this change.

felixxm commented on django/db/models/sql/query.py:
```diff
@@ -1619,6 +1625,8 @@ def names_to_path(self, names, opts, allow_many=True, fail_on_missing=False):
             filtered_relation = None
             try:
+                if opts is None:
+                    raise FieldDoesNotExist
```
This is only tested in `constraints.tests.CheckConstraintTests.test_validate` and `constraints.tests.UniqueConstraintTests.test_validate_expression_condition`. Can we add a regression test in the `FromEmptyQuery`?

felixxm commented on tests/queries/test_query.py:
```diff
@@ -158,3 +159,31 @@ def test_repr(self):
+        query.add_annotation(RawSQL("%s IS NULL", (None,), BooleanField()), "_check")
+        compiler = query.get_compiler(using="default")
+        result = compiler.execute_sql(SINGLE)
+        self.assertTrue(result[0])
```
Use `assertIs(…, True/False)` for testing boolean values, rather than `assertTrue()` and `assertFalse()`, so you can check the actual boolean value, not the truthiness of the expression, as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

felixxm reviewed (commented):
@Gagaro Thanks for all your efforts :star: I reviewed the first commit. Please move it to a separate PR. I'd like to merge it in advance.

felixxm commented:
@Gagaro Thanks :+1: I moved the first commit to a separate PR, see #15512.

felixxm commented:
@Gagaro I created the second PR with a small cleanup from this patch, see #15515.

felixxm reviewed (commented):
@Gagaro I rebased from the `main` branch. Unfortunately, many tests fail on Oracle because it doesn't support boolean expressions in the `SELECT` clause, so we should wrap it with `CASE WHEN` expression, as in

https://github.com/django/django/blob/1ea7e3157d1f9b4db71e768d75ea57e47dbd49f9/django/db/models/expressions.py#L1360-L1366

Another idea is to move conditions to the `WHERE` clause (instead of the `SELECT`) and check if query returns any rows, e.g.:
- on Oracle:
  ```sql
  SELECT 1
  FROM DUAL
  WHERE LOWER("X") = "sth"
  ```
  instead of
  ```sql
  SELECT CASE WHEN LOWER("X") = "sth" THEN 1 ELSE 0 END
  FROM DUAL
  ```
- on PosgreSQL
  ```sql
  SELECT TRUE
  WHERE LOWER("X") = "sth"
  ```
  instead of
  ```sql
  SELECT LOWER("X") = "sth"
  ```

(\cc @charettes)

charettes commented:
@felixxm no strong opinion here. It looks like one way or the other changes can be encapsulated in `Q.check` or `ExpressionWrapper.select_format`.

I tried moving `Exists.select_format` to `BaseExpression.select_format` instead when `isinstance(self.output_field, BooleanField)` but it has the unfortunate effect of wrapping `Col` as well.

I wonder if switching  `supports_boolean_expr_in_select_clause` to a backend specific method `select_expression_fomat(expression: Expression) -> (sql: str, params: tuple)` might make things a bit more uniform here.

felixxm commented:
Switching to the `WHERE` clause should simplify generated query, now it causes multiple layers of `CASE WHEN` wraps, e.g.
```sql
SELECT CASE WHEN (
    NOT (:arg0 IS NULL) OR
    CASE
        WHEN NOT EXISTS(
            SELECT (1) AS "A"
            FROM "CONSTRAINTS_UNIQUECONSTRAIFD7C" U0
            WHERE LOWER(U0."NAME") = (LOWER(:arg1))
            FETCH FIRST 1 ROWS ONLY
        ) THEN :arg2
        ELSE :arg3
        END = :arg2
) THEN 1 ELSE 0 AS "_CHECK"
FROM DUAL
```
instead of
```sql
SELECT 1
FROM DUAL
WHERE
    NOT (:arg0 IS NULL) OR 
    EXISTS(
        SELECT (1) AS "A"
        FROM "CONSTRAINTS_UNIQUECONSTRAIFD7C" U0
        WHERE LOWER(U0."NAME") = (LOWER(:arg1))
        FETCH FIRST 1 ROWS ONLY
    )
```

Gagaro commented on docs/releases/4.1.txt:
```diff
@@ -268,6 +268,10 @@ Models
   in order to reduce the number of failed requests, e.g. after database server
   restart.
 
+* New :meth:`BaseConstraint.validate(instance, exclude=None, using=DEFAULT_DB_ALIAS) <.BaseConstraint.validate>`
```
Shouldn't this be `are checked during ``full_clean()``.`?

felixxm replied:
Typo, `are now checked ...`.

Gagaro commented on docs/ref/models/constraints.txt:
```diff
@@ -201,3 +226,13 @@ For example::
+
+Message used for ``ValidationError`` raised by ``validate()``.
+``{name}`` can be specified in the message and is replaced by the constraint
+name.
```
I'm not sure to understand what you mean exactly here. Should we just move the `message` part to the `BaseConstraint` part and let the name be in both `CheckConstraint` and `UniqueConstraint` for now? Or can we move the name as well now?

Gagaro commented:
Thanks for all the follow ups :+1: .

I fixed the doc and created the PR for `BaseConstraint` (#15518).

I'll take a look at the other issue when I'll have more time, which will probably be next week.

Gagaro commented:
I'm not sure what's up with MySQL, it's working locally. Is there a way to know the exact version used in the CI?

felixxm commented:
> I'm not sure what's up with MySQL, it's working locally. Is there a way to know the exact version used in the CI?

See https://code.djangoproject.com/wiki/CI.

Gagaro commented:
MySQL 5.7 needs a `FROM` clause when there is a `WHERE` clause.

What is strange is that it is not documented as such though: https://dev.mysql.com/doc/refman/5.7/en/select.html

I found this old stackoverflow question which talks about it (for 5.5): https://stackoverflow.com/questions/27150445/select-where-without-from-throws-syntax-error

I added `bare_select_suffix = " FROM DUAL"` to the MySQL `DatabaseFeatures`.

felixxm commented:
buildbot, test on oracle.

Gagaro commented:
I rebased on main and removed the MySQL fix.

felixxm commented on django/db/models/constraints.py:
```diff
@@ -265,3 +294,63 @@ def deconstruct(self):
+                    for constraint in constraints:
+                        if type(constraint) is type(self):
+                            raise ValidationError(
+                                instance.unique_error_message(model, self.fields)
```
This blocks overriding an error message via `UniqueConstraint.message` in most of cases and should at least be documented.

felixxm commented on django/db/models/base.py:
```diff
@@ -1394,10 +1384,32 @@ def unique_error_message(self, model_class, unique_check):
         """
-        Call clean_fields(), clean(), and validate_unique() on the model.
-        Raise a ValidationError for any errors that occur.
+        Call clean_fields(), clean(), validate_unique(), and validate_constraints()
```
Wrap at 79 chars.

felixxm commented on django/contrib/postgres/constraints.py:
```diff
@@ -197,3 +200,47 @@ def __repr__(self):
+            if queryset.exists():
+                raise ValidationError(self.get_message())
+        else:
+            try:
+                if not (~self.condition | ~Exists(queryset)).check(replacement_map):
+                    raise ValidationError(self.get_message())
+            except FieldError:
+                return None
```
`FieldError` is untested. Do we need it? It looks unnecessary.

felixxm commented on django/db/models/constraints.py:
```diff
@@ -25,6 +34,12 @@ def create_sql(self, model, schema_editor):
         raise NotImplementedError("This method must be implemented by a subclass.")
 
+    def validate(self, model, instance, exclude=None, using=DEFAULT_DB_ALIAS):
+        raise NotImplementedError("This method must be implemented by a subclass.")
```
This is untested.

felixxm commented on tests/postgres_tests/test_constraints.py:
```diff
@@ -731,6 +732,115 @@ def test_range_adjacent(self):
+        )
+        constraint.validate(HotelReservation, reservation)
+        # Overlap dates.
+        with self.assertRaises(ValidationError):
```
```suggestion
        with self.assertRaisesMessage(ValidationError, msg):
```

felixxm reviewed (commented):
@Gagaro Thanks :+1: I left more comments.

Gagaro commented:
Github actions seems to not have ran. I pushed again to try to have them run.

felixxm commented on django/db/models/constraints.py:
```diff
@@ -1,16 +1,25 @@
 
 class BaseConstraint:
-    def __init__(self, name):
+    violation_error_message = _("Constraint {name} is violated.")
```
I would add quotes:
```suggestion
    violation_error_message = _("Constraint '{name}' is violated.")
```

felixxm reviewed (commented):
@Gagaro Thanks for updates :+1:

Unfortunately, I noticed another issue. `CheckConstraint` supports any conditional expression not only `Q()` (see e9a0e1d4f6ecfc4227acff74e1f56f288a0b30aa) but they don't implement the `check()` method, so `validate()` crashes for them:
```diff
diff --git a/tests/constraints/tests.py b/tests/constraints/tests.py
index 078da6902d..7ef621e027 100644
--- a/tests/constraints/tests.py
+++ b/tests/constraints/tests.py
@@ -177,6 +177,33 @@ class CheckConstraintTests(TestCase):
             exclude={"discounted_price", "price"},
         )
 
+    def test_validate_boolean_expressions(self):
+        tests = [
+            models.CheckConstraint(
+                check=models.expressions.RawSQL(
+                    "price < %s OR price > %s",
+                    (500, 500),
+                    output_field=models.BooleanField(),
+                ),
+                name="price_neq_500_raw",
+            ),
+            models.CheckConstraint(
+                check=models.expressions.ExpressionWrapper(
+                    models.Q(price__gt=500) | models.Q(price__lt=500),
+                    output_field=models.BooleanField(),
+                ),
+                name="price_neq_500_wrap",
+            ),
+        ]
+        for constraint in tests:
+            with self.subTest(constraint=constraint):
+                constraint.validate(Product, Product(price=500, discounted_price=5))
+                msg = f"Constraint {constraint.name} is violated."
+                with self.assertRaisesMessage(ValidationErrorm, msg):
+                    constraint.validate(Product, Product(price=501, discounted_price=5))
+                with self.assertRaisesMessage(ValidationErrorm, msg):
+                    constraint.validate(Product, Product(price=499, discounted_price=5))
+
 
 class UniqueConstraintTests(TestCase):
     @classmethod

```

Gagaro commented:
So it seems to work only on sqlite and mysql. We can get the value in postgres with something like :

```
SELECT toto FROM (SELECT 1 as toto) AS t WHERE toto = 1;
```

Gagaro commented:
It also works on SQLite and MySQL.

For this we would need https://code.djangoproject.com/ticket/24462 (it could also work with a CTE https://code.djangoproject.com/ticket/28919).

Do you think of another way to have the values for RawSQL?

Gagaro commented:
Thinking about it, maybe we should just not do the check if there is a `RawSQL` in the `Q` and document it as such. It would raise an integrity error, like we currently have.

Then, when one of the above is implemented, we might use that to check on `Q` with `RawSQL`.

What do you think about it?

charettes reviewed (commented):
I'm not sure we can get this feature to work at all for `RawSQL` and any form of _baked_ column references. Getting the right SQL is one thing but determining if a check using _baked_ column reference should be run of `exclude_fields` is another (e.g. what happens if `exclude_fields={'price'}` in your example?).

With that in mind I think we need `Q.check` to catch database errors (or maybe only the ones for a missing field reference) and turn them into a form of runtime warning including the failed SQL and error and return `True`.

Adding a system check to detect the usage of `RawSQL` and warn users about it is also something we should consider?

Gagaro commented on django/db/models/query_utils.py:
```diff
@@ -95,6 +96,36 @@ def resolve_expression(
+        query = Query(None)
+        # `select` needs to be True so that RawSQL expressions have access
+        # to the values
+        select = any((isinstance(expr, RawSQL) for expr in self.children))
```
Indeed, I meant `self.flatten` instead of `self.children`. Which made me realize my `Q.flatten()` implementation wasn't actually correct. I pushed a fix for it.

Gagaro commented:
> I'm not sure we can get this feature to work at all for `RawSQL` and any form of _baked_ column references. Getting the right SQL is one thing but determining if a check using _baked_ column reference should skipped due to `exclude_fields` is another (e.g. what happens if `exclude_fields={'price'}` in your example?).

Indeed I forgot about the exclusion. In that case, the query will return an error, that may be a way to handle that.

> With that in mind I think we need `Q.check` to catch database errors (or maybe only the ones for a missing field reference e.g. [42703 for Postgres](https://www.postgresql.org/docs/14/errcodes-appendix.html)) and turn them into a form of runtime warning including the failed SQL and error and return `True`.

Should we really show a warning in this case? This would be a warning the user can not fix, as this is a limitation in the ORM and not something they could act on.

> Adding a system check to detect the usage of `RawSQL` and warn users about it is also something we should consider?

Yes we could do that too, at the info level?

charettes commented:
> Should we really show a warning in this case? This would be a warning the user can not fix, as this is a limitation in the ORM and not something they could act on.

Most of the usage of `RawSQL`, particularly in the case of the subset of SQL allowed in constraints, can be expressed using a proper expression that doesn't _bake_ column references but take them as `F` that resolves to `Col` so definition something users can act on. Messages emitted by `warnings.warn` are also silenceable using advanced filtering.

> Yes we could do that too, at the info level?

It feels like it would be more of a warning since it's something a user can address and could be the cause of unexpected integrity errors.

charettes commented:
I'll add that the `RawSQL` usage detection only goes so far; nothing prevents users from writing expressions that bake column references.

e.g.


```python
Func(50, template="price > %(expressions)s OR price < %(expressions)s")
```

Gagaro commented:
All right then. I had in mind that someone using `RawSQL` would only do it because they don't have the choice.

Gagaro commented:
I added the warning in `Q.check()` and the system check.

The check is showing up in the tests because we have a model with a `CheckConstraint` containing a `RawSQL`. I'm not sure if we can suppress it?

Gagaro commented:
Rebased on main.

What's next for this PR?

felixxm commented:
> What's next for this PR?

All good :+1: . Unfortunately, you have to be patient, but don't worry it's on my list for 4.1.

Gagaro commented:
> > What's next for this PR?
> 
> All good +1 . Unfortunately, you have to be patient, but don't worry it's on my list for 4.1.

Awesome, thanks. I asked in case I had more things to do :smiley:  .

felixxm commented on django/db/models/query_utils.py:
```diff
@@ -95,6 +99,39 @@ def resolve_expression(
+        for child in self.children:
+            if isinstance(child, tuple):
+                child = child[1]
+            if hasattr(child, "flatten"):
+                yield from child.flatten()
```
This should yield `child` when it doesn't implement `flatten()`.

felixxm reviewed (commented):
@Gagaro Thanks :+1: I will try to review it in detail and maybe merge some pieces beforehand. Please don't rebase or force-push to this branch anymore as it will make reviewing difficult for me.

I moved adding `Q.flatten()` to a separate PR, see #15654.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
I moved adding `Q.check()` with small edits to the #15659.

felixxm commented:
> I added the warning in `Q.check()` and the system check.
> 
> The check is showing up in the tests because we have a model with a `CheckConstraint` containing a `RawSQL`. I'm not sure if we can suppress it?

We can move these constraints to `migrations` tests to avoid warnings, see #15660.

felixxm reviewed (commented):
@Gagaro I pushed the first set of edits.

felixxm commented on django/core/checks/model_checks.py:
```diff
@@ -38,6 +42,18 @@ def check_all_models(app_configs=None, **kwargs):
+                        obj=model,
+                        id="models.W045",
+                    )
+                )
```
This should be moved to `Model._check_constraints()`.

felixxm commented:
buildbot, test on oracle.

felixxm commented on django/db/models/base.py:
```diff
@@ -1410,10 +1411,33 @@ def unique_error_message(self, model_class, unique_check):
+            for constraint in model_constraints:
+                try:
+                    constraint.validate(model_class, self, exclude=exclude, using=using)
+                except ValidationError as e:
+                    errors = e.update_error_dict(errors)
```
As far as I'm aware, for backward compatibility we should assigned errors from `UniqueConstraint`s with a single field to this field :thinking:

felixxm commented:
I finished and squashed my edits, IMO it's ready :fire: 

@Gagaro Thanks for all your efforts :rocket: :star2:

Gagaro commented:
Thanks a lot for the reviews and the support everyone :heart: .

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on selenium.

carltongibson commented on docs/releases/4.1.txt:
```diff
@@ -65,6 +65,15 @@ advantage of developments in the ORM's asynchronous support as it evolves.
+:class:`unique <django.db.models.UniqueConstraint>`, and :class:`exclusion
+<django.contrib.postgres.constraints.ExclusionConstraint>` constraints defined
+in the :attr:`Meta.constraints <django.db.models.Options.constraints>` option
+are now checked during the :ref:`model validation <validating-objects>`.
```
```suggestion
are now checked during :ref:`model validation <validating-objects>`.
```

carltongibson commented on docs/ref/models/constraints.txt:
```diff
@@ -203,3 +232,21 @@ For example::
+
+.. attribute:: UniqueConstraint.violation_error_message
+
+The error message used when ``ValidationError`` is raised during the
+:ref:`model validation <validating-objects>`. Defaults to
```
```suggestion
The error message used when ``ValidationError`` is raised during
:ref:`model validation <validating-objects>`. Defaults to
```

carltongibson commented on docs/ref/models/constraints.txt:
```diff
@@ -60,10 +58,37 @@ All constraints have the following parameters in common:
+
+.. attribute:: BaseConstraint.violation_error_message
+
+The error message used when ``ValidationError`` is raised during the
+:ref:`model validation <validating-objects>`. Defaults to
```
```suggestion
The error message used when ``ValidationError`` is raised during
:ref:`model validation <validating-objects>`. Defaults to
```

carltongibson commented on docs/ref/models/constraints.txt:
```diff
@@ -31,24 +31,22 @@ option.
+
+.. versionchanged:: 4.1
+
+    In older versions, constraints were not checked during the model
+    validation.
```
```suggestion
    In older versions, constraints were not checked during model
    validation.
```

carltongibson commented on docs/ref/contrib/postgres/constraints.txt:
```diff
@@ -165,6 +173,15 @@ creates an exclusion constraint on ``circle`` using ``circle_ops``.
+
+.. versionadded:: 4.1
+
+The error message used when ``ValidationError`` is raised during the
+:ref:`model validation <validating-objects>`. Defaults to
```
```suggestion
The error message used when ``ValidationError`` is raised during
:ref:`model validation <validating-objects>`. Defaults to
```

carltongibson commented on docs/ref/contrib/postgres/constraints.txt:
```diff
@@ -27,6 +27,14 @@ PostgreSQL supports additional data integrity constraints available from the
+
+    .. versionchanged:: 4.1
+
+        In older versions, exclusion constraints were not checked during the
+        model validation.
```
```suggestion
        In older versions, exclusion constraints were not checked during
        model validation.
```

carltongibson reviewed (commented):
Looks great. Thanks @Gagaro and @felixxm. 

I looks like a spare `the` in some of the docs changes… An alternate might by `during validation of the model` if emphasising the instance is what you wanted. 🤔

felixxm commented:
@carltongibson Thanks :+1:

felixxm commented:
buildbot, test on selenium.

blueyed commented on django/db/models/constraints.py:
```diff
@@ -265,3 +289,61 @@ def deconstruct(self):
+                ):
+                    raise ValidationError(self.get_violation_error_message())
+            except FieldError:
+                pass
```
I've noticed that a field used in a condition might be excluded/missing in the `against` map, because it is excluded from the form (an internal status field).
The resulting `FieldError` is swallowed then.
Is this expected?  Why should anything get excluded from `against` here (and likely other places from this patch)?

charettes replied:
This is [the documented behaviour for forms](https://docs.djangoproject.com/en/4.1/topics/forms/modelforms/#interaction-with-model-validation)

> If you have excluded any model fields, validation will not be run on those fields. See the [form validation](https://docs.djangoproject.com/en/4.1/ref/forms/validation/) documentation for more on how field cleaning and validation work.

I guess the [form validation](https://docs.djangoproject.com/en/4.1/ref/forms/validation/) section could go into more details on the subject but this is how `unique_together` validation has been working forever

https://github.com/django/django/blob/08688bd7dd8dfa218ecff03d6c94a753b1be8e59/django/db/models/base.py#L1235-L1242

blueyed replied:
The point is that I am not using `exclude` myself, and it is good that `status` is excluded from checks itself.
But the constraint is for `name`, and `status` is only used in the condition there:
```
    class Meta(…):
        constraints = [
            UniqueConstraint(fields=("name",), condition=Q(status="draft"), name="uniq_name_draft"),
        ]
 ```

blueyed replied:
> It just happens to pretty straightforward here as you can directly call `Constraint.validate` without the `exclude` on the constraint you are interested in validating.

That's a suitable workaround, but I feel like it should not be necessary.
FWIW before Django 4.1 where this feature was added I added manual validation already, since there constraints with conditions where just skipped.
